### PR TITLE
feat: support download as image/pdf

### PR DESCRIPTION
**Pull request by quanho**
This PR is to support download as image/pdf for pagination table

### CODE REVIEW

File: src/table/hooks/use-snapshot.ts [hunk 1]
```diff
@@ -35,10 +36,28 @@ const useSnapshot = ({ layout, viewService, model, rootElement }: UseSnapshotPro
...
+        const totalsPosition = getTotalPosition(layout);
+        const { visibleRowStartIndex = -1, visibleRowEndIndex = -1 } = findVisibleRows(rootElement, totalsPosition);
+        return {
+          scrollLeft: viewService.scrollLeft,
+          visibleTop: viewService.qTop + visibleRowStartIndex,
+          visibleHeight: visibleRowEndIndex < 0 ? 0 : visibleRowEndIndex - visibleRowStartIndex + 1,
+          rowsPerPage: viewService.rowsPerPage,
+          page: viewService.page,
+        };
+      },
```
Comment (haxxmaxx) [inline src/table/hooks/use-snapshot.ts:57]: This looks like duplicate code. Can you break it out?

File: src/table/hooks/use-view-service.ts [hunk 2]
```diff
@@ -1,16 +1,21 @@
...
+const createViewService = (viewState: ViewState, snapshotData?: SnapshotData): ViewService => {
   return {
     qLeft: 0,
     qTop: 0,
     qWidth: 0,
     qHeight: 0,
-    scrollLeft: 0,
+    visibleTop: viewState?.visibleTop,
+    visibleHeight: viewState?.visibleHeight,
+    scrollLeft: snapshotData?.content?.scrollLeft ?? (viewState?.scrollLeft || 0),
```
Comment (haxxmaxx) [inline src/table/hooks/use-view-service.ts:12]: would be nice with consistency here, does 
```suggestion
    scrollLeft: snapshotData?.content?.scrollLeft ?? viewState?.scrollLeft ?? 0,
```
work here?

File: src/table/pagination-table/components/TableWrapper.tsx [hunk 3]
```diff
@@ -119,7 +119,7 @@ function TableWrapper(props: TableWrapperProps) {
       <StyledTableContainer
         ref={tableContainerRef}
         className="sn-table-container"
-        fullHeight={footerContainer || constraints.active || !paginationNeeded} // the footerContainer always wants height: 100%
+        fullHeight={footerContainer || !paginationNeeded} // the footerContainer always wants height: 100%
```
[the remark this answers is not in the document]
Comment (quanho) [reply src/table/pagination-table/components/TableWrapper.tsx:122]: Sorry, that only for testing! Will remove it!

File: src/index.ts [hunk 4]
```diff
@@ -46,7 +47,8 @@ export default function supernova(env: Galaxy) {
       const selectionsAPI = useSelections() as ExtendedSelectionAPI | undefined; // undefined when taking snapshot
       const keyboard = useKeyboard();
       const rect = useRect();
-      const viewService = useViewService();
+      const { viewState } = useOptions() as UseOptions;
+      const viewService = useViewService(viewState, layout.snapshotData);
```
[the remark this answers is not in the document]
Comment (quanho) [reply src/index.ts:51]: Updated!

File: src/table/hooks/use-snapshot.ts (hunk 1 above)
Comment (quanho) [reply src/table/hooks/use-snapshot.ts:57]: Yes, it partly duplicate. Both onTakeSnapshot and getViewState need to return scrollLeft, rowsPerPage and page, but getViewState need visibleTop and visibleHeight for fetching correct data rows while onTakeSnapshot does not need them.

Comment (quanho) [reply src/table/hooks/use-snapshot.ts:57]: Updated!

File: src/table/hooks/use-snapshot.ts [hunk 5]
```diff
@@ -1,4 +1,5 @@
-import { onTakeSnapshot } from '@nebula.js/stardust';
+// @ts-ignore ignore useImperativeHandle
```
Comment (LiKang6688) [inline src/table/hooks/use-snapshot.ts:1]: ts-expect-error is a better alternative in this case

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#ts-ignore-or-ts-expect-error

**Review by haxxmaxx — APPROVED** [review]
LGTM

File: src/table/hooks/use-view-service.ts [hunk 6]
```diff
@@ -1,16 +1,23 @@
-import { useMemo } from '@nebula.js/stardust';
-import type { ViewService } from '../../types';
+import { useMemo, useOptions } from '@nebula.js/stardust';
+import type { SnapshotData, ViewService, ViewState, UseOptions } from '../../types';
 
-const createViewService = (): ViewService => {
+const createViewService = (viewState: ViewState, snapshotData?: SnapshotData): ViewService => {
```
Comment (LiKang6688) [inline src/table/hooks/use-view-service.ts:4]: `snapshotData: SnapshotData | undefined`  is more precise in this case, as snapshotData is not really optional but could be undefined

when `snapshotData?: SnapshotData`, `createViewService(viewState)` will not complained by ts

File: src/table/hooks/use-view-service.ts [hunk 7]
```diff
@@ -1,16 +1,23 @@
...
+    visibleTop: viewState?.visibleTop,
+    visibleHeight: viewState?.visibleHeight,
+    scrollLeft: snapshotData?.content?.scrollLeft ?? viewState?.scrollLeft ?? 0,
+    rowsPerPage: snapshotData?.content?.rowsPerPage ?? viewState?.rowsPerPage,
+    page: snapshotData?.content?.page ?? viewState?.page,
   };
 };
 
-const useViewService = (): ViewService => useMemo(() => createViewService(), []);
+const useViewService = (snapshotData?: SnapshotData): ViewService => {
```
Comment (LiKang6688) [inline src/table/hooks/use-view-service.ts:18]: ditto

**Pull request merged**